### PR TITLE
[ESIMD] Do simd<T, N>* to <N x T> arg/ret type conv when possible.

### PR DESCRIPTION
**Pull request by kbobrovs**
Today, `simd<T, N> foo(simd<T, N> x)` is codegenerated by clang as
   `void foo(simd<T, N>* sret(simd<T, N>) %res, simd<T, N>* %x)`
for the SPIRV target (unless `__regcall` is specified), which is then converted to
  `void foo(<N x T>* sret(<N x T>) %res, <N x T>* %x)`
in the LowerESIMDVecArg, then to
  `<N x T> foo(<N x T> %x)`
in the VC BE.

With the opaque pointers this becomes impossible, and the optimization must happen in the "ESIMD FE". This patch implements it.

It also changes `lowerEsimdConstructs` in sycl-post-link.cpp to use new pass manager to
avoid the old pass manager-related boiler plate code in new ESIMD transformations.

Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

**Review by asudarsa — APPROVED** [review]
sycl-post-link changes look good to me. Thanks

File: llvm/lib/SYCLLowerIR/ESIMD/ESIMDOptimizeVecArgCallConv.cpp [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+  NewF->setName(Name);
+  return true;
+}
+
+namespace llvm {
+
+PreservedAnalyses
+ESIMDOptimizeVecArgCallConvPass::run(Module &M, ModuleAnalysisManager &MAM) {
+  bool Modified = false;
+  SmallPtrSet<Function *, 32> Visited;
```
Comment (v-klochkov) [inline llvm/lib/SYCLLowerIR/ESIMD/ESIMDOptimizeVecArgCallConv.cpp:479]: nit: THis is unused, but can be removed in separate PR.

**Review by v-klochkov — APPROVED** [review]
It also would be nice to have a special test for the case when simd is passed by reference and is written in the callee. Also, can be done in separate PR.

**Comment by kbobrovs** [comment]
> It also would be nice to have a special test for the case when simd is passed by reference and is written in the callee. Also, can be done in separate PR.

This is a very good point. I will add as a follow-up.

**Comment by pvchupin** [comment]
@kbobrovs, please fix post-commit https://github.com/intel/llvm/actions/runs/3100490530

**Comment by asudarsa** [comment]
Look like we need to add 'Analysis' to the LLVM_LINK_COMPONENTS' in sycl-post-link/CMakeLists.txt ?

**Comment by kbobrovs** [comment]
Yes, will do

**Comment by kbobrovs** [comment]
Fix https://github.com/intel/llvm/pull/6844
